### PR TITLE
Add Support for PCT2075

### PR DESCRIPTION
**Pull request by dkhayes117**
Added full support for the PCT2075 sensor, a pin for pin replacement for the LM75.  The PCT2075 has 11-bit ADC resolution where the LM75 is only 9-bit.  This is handled to where all 11-bits are used where applicable.  

Added support for the PCT2075's extra pointer register T_idle which sets the sample rate of the sensor in increments of 100 milliseconds, only bits [4:0] of the byte are used for a max period of 3100ms or 3.1s.  Any writes to bits [7:5] are ignored. Default is 100ms, writing zero to the register defers to the default value.  Sample rate value is register value * 100ms.

Fix bug and optimize the temperature conversion functions.  The bug wrote 0b0000_0001 to the LSB for 0.5 degree increments, however the correct implementation is 0x1000_0000 for the LSB.   

Relevant tests and doc examples added for new functionality.  

PCT2075 implementation uses a different constructor, `new_pct2075()`, to differentiate extra functionality from the standard LM75 instance using `new()`.

Cargo build passing, cargo test passing :)

### CODE REVIEW

File: src/conversion.rs [hunk 1]
```diff
@@ -1,115 +1,114 @@
...
-    }
-    if temp - f32::from(temp as i8) > 0.499 {
-        (msb as u8, 1)
-    } else {
-        (msb as u8, 0)
-    }
+pub fn convert_temp_to_register(temp: f32, mask: Resolution) -> (u8, u8) {
+    let int = (temp / 0.125) as i16 as u16;
+    let binary = int << 5;
+    let bytes = binary.to_be_bytes();
```
Comment (eldruin) [inline src/conversion.rs]: `to_be_bytes` was only added in Rust 1.32.0 but I would like to keep the current MSRV.
Can we replace this with some masking and shifting?

File: src/conversion.rs [hunk 2]
```diff
@@ -1,115 +1,114 @@
...
+    (bytes[0], bytes[1] & mask as u8)
+}
+
+pub fn convert_sample_rate_from_register(byte: u8) -> u16 {
+    // Bits [4:0] hold sample rate value
+    ((byte as u16) & 0x1F) * 100
```
Comment (eldruin) [inline src/conversion.rs]: Can we put the `0x1F` constant somewhere? Probably in `BitFlags`.

File: src/conversion.rs [hunk 3]
```diff
@@ -1,115 +1,114 @@
...
-        assert_near!(convert_temp_from_register(0b0000_0010, 0b0101_1010), 2.0);
-        assert_near!(convert_temp_from_register(0b0000_0100, 0b0101_1010), 4.0);
-        assert_near!(convert_temp_from_register(0b0000_1000, 0b0101_1010), 8.0);
-        assert_near!(convert_temp_from_register(0b0001_0000, 0b0101_1010), 16.0);
-        assert_near!(convert_temp_from_register(0b0010_0000, 0b0101_1010), 32.0);
-        assert_near!(convert_temp_from_register(0b0100_0000, 0b0101_1010), 64.0);
-        assert_near!(convert_temp_from_register(0b0100_1011, 0b0101_1010), 75.0);
-        assert_near!(convert_temp_from_register(0b0101_0000, 0b0101_1010), 80.0);
-        assert_near!(convert_temp_from_register(0b0111_1111, 0b1101_1010), 127.5);
+        assert_eq!(convert_temp_from_register(0b0111_1101, 0b0101_1010, Mask9bit), 125.0);
```
Comment (eldruin) [inline src/conversion.rs]: It would be better to keep the `assert_near!` here. Albeit with a much smaller epsilon.

File: src/device_impl.rs [hunk 4]
```diff
@@ -113,12 +129,26 @@ where
...
         }
-        let (msb, lsb) = conversion::convert_temp_to_register(temperature);
+        let (msb, lsb) = conversion::convert_temp_to_register(temperature, self.resolution);
         self.i2c
             .write(self.address, &[Register::T_HYST, msb, lsb])
             .map_err(Error::I2C)
     }
 
+    /// Set the sensor sample rate period (100ms increments).
+    pub fn set_sample_rate(&mut self, period: u16) -> Result<(), Error<E>> {
```
Comment (eldruin) [inline src/device_impl.rs]:
```suggestion
    /// Set the sensor sample rate period in milliseconds (in 100ms increments).
    ///
    /// For values outside of the range `[100 - 3100]` or those not multiple of 100, 
    /// `Error::InvalidInputData` will be returned.
    pub fn set_sample_rate(&mut self, period_ms: u16) -> Result<(), Error<E>> {
```

File: src/device_impl.rs [hunk 5]
```diff
@@ -113,12 +129,26 @@ where
...
+        if self.sample_rate.bits.is_none() {
+            return Err(Error::InvalidRegister);
```
Comment (eldruin) [inline src/device_impl.rs]: With a marker type instead we could implement `set_sample_rate` only for capable devices and avoid having this error.

File: src/lib.rs [hunk 6]
```diff
@@ -136,22 +147,33 @@
...
 //! ```
+//! ### Set the Sample Rate on the PCT2075 only
+//!
+//! ```no_run
+//! use linux_embedded_hal::I2cdev;
+//! use lm75::{Lm75, Address};
+//!
+//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
+//! let mut sensor = Lm75::new_pct2075(dev, Address::default());
+//! let period = 1500; //in milliseconds, max = 3100, default 100
```
Comment (eldruin) [inline src/lib.rs]:
```suggestion
//! let period = 1500; // in milliseconds, max = 3100, default 100
```

File: src/lib.rs [hunk 7]
```diff
@@ -76,58 +76,69 @@
...
+//! use lm75::{Lm75, Address};
 //!
 //! let dev = I2cdev::new("/dev/i2c-1").unwrap();
 //! let (a2, a1, a0) = (false, false, true);
-//! let address = SlaveAddr::Alternative(a2, a1, a0);
+//! let address = Address::from((a2,a1,a0));
 //! let mut sensor = Lm75::new(dev, address);
 //! ```
 //!
+//! ### Provide a custom address for PCT2075 only
```
Comment (eldruin) [inline src/lib.rs]:
```suggestion
//! ### Provide a full custom address
```

File: src/lib.rs [hunk 8]
```diff
@@ -76,58 +76,69 @@
...
+//!
+//! ```no_run
+//! use linux_embedded_hal::I2cdev;
+//! use lm75::{Lm75, Address};
+//!
+//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
+//! let all_pins_floating = 0x37; // 0b011_0111
```
Comment (eldruin) [inline src/lib.rs]:
```suggestion
//! let all_pins_floating = 0x37; // PCT2075 supports many addresses
```

File: src/lib.rs [hunk 9]
```diff
@@ -136,22 +147,33 @@
...
-//! use lm75::{Lm75, SlaveAddr};
+//! use lm75::{Lm75, Address};
 //!
 //! let dev = I2cdev::new("/dev/i2c-1").unwrap();
-//! let mut sensor = Lm75::new(dev, SlaveAddr::default());
+//! let mut sensor = Lm75::new(dev, Address::default());
 //! let temp_celsius = 40.0;
 //! sensor.set_hysteresis_temperature(temp_celsius).unwrap();
 //! ```
+//! ### Set the Sample Rate on the PCT2075 only
```
Comment (eldruin) [inline src/lib.rs]:
```suggestion
//! ### Set the sample rate (PCT2075 only)
```

File: src/lib.rs [hunk 10]
```diff
@@ -228,6 +255,20 @@ pub enum OsMode {
     Interrupt,
 }
 
+#[derive(Debug, Clone, Copy)]
+/// Device Resolution
+pub enum Resolution {
```
Comment (eldruin) [inline src/lib.rs]:
```suggestion
pub (crate) enum Resolution {
```
This does not need to be public, right?

File: src/lib.rs [hunk 11]
```diff
@@ -228,6 +255,20 @@ pub enum OsMode {
...
+    /// 9bit has 0.5 resolution, 11bit has 0.125
+    /// Masks the LSB only
```
Comment (eldruin) [inline src/lib.rs]:
```suggestion
    /// Sensors with 9-bit resolution (0.5ºC steps)
```

File: src/lib.rs [hunk 12]
```diff
@@ -228,6 +255,20 @@ pub enum OsMode {
...
+    Mask9bit = 0b1000_0000,
+    /// Sensors with 11-bit resolution (PCT2075)
```
Comment (eldruin) [inline src/lib.rs]:
```suggestion
    /// Sensors with 11-bit resolution (PCT2075) (0.125ºC steps)
```

File: src/lib.rs [hunk 13]
```diff
@@ -263,6 +317,10 @@ pub struct Lm75<I2C> {
     address: u8,
     /// Configuration register status.
     config: Config,
+    /// Device Resolution
+    resolution: Resolution,
+    /// T-Idle Register Contents
+    sample_rate: SampleRate,
```
Comment (eldruin) [inline src/lib.rs]: We can remove both when we use marker types.

**Review by eldruin — COMMENTED** [review]
Thank you for your work!

File: src/conversion.rs [hunk 14]
```diff
@@ -1,115 +1,89 @@
...
-    // msb is stored as two's complement so we can just do this:
-    let value = msb as i8;
-    if (lsb >> 7) != 0 {
-        f32::from(value) + 0.5
-    } else {
-        f32::from(value)
-    }
+    // msb is stored as two's complement
+    let msb = msb as i8 as f32;
+    let decimal = ((lsb & mask as u8) >> 5) as f32 * 0.125;
```
Comment (eldruin) [inline src/conversion.rs]: Somehow the mask has been removed

**Comment by dkhayes117** [comment]
Okay, I think I've got it :+1: .  Builds with all tests passing.

**Comment by dkhayes117** [comment]
Added some missing info to the README.md and the doc comments in lib.rs.  Should be all there now :)

File: src/lib.rs [hunk 15]
```diff
@@ -168,30 +194,28 @@ pub enum Error<E> {
...
-    /// Alternative slave address providing bit values for A2, A1 and A0
-    Alternative(bool, bool, bool),
-}
+/// I2C device address
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Address(pub(crate) u8);
 
-impl Default for SlaveAddr {
-    /// Default slave address
+/// Default device
```
Comment (eldruin) [inline src/lib.rs]:
```suggestion
/// Default address
```

**Review by eldruin — COMMENTED** [review]
Looks good! Could you fix the clippy warnings before the merge?

**Review by eldruin — COMMENTED** [review]
Could you add an integration test reading the temperature for PCT2075?
Something similar to this but for a value which implies the right definition is used like 24.125:
```rust
#[test]
fn can_read_temperature() {
    let mut sensor = new_pct2075(&[I2cTrans::write_read(
        ADDR,
        vec![Register::TEMPERATURE],
        vec![0b1110_0111, 0b1010_0101], // -24.5
    )]);
    let temp = sensor.read_temperature().unwrap();
    assert!(-24.4 > temp);
    assert!(-24.6 < temp);
    destroy(sensor);
}
```
That would prove that the adequate bit mask is used when reading.

**Comment by dkhayes117** [comment]
Sure, I will add it this evening.  Appreciate it!

**Review by eldruin — APPROVED** [review]
Great work, thank you very much!

**Pull request merged**